File: python/002_addTwoNumbers/addTwoNumbers.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        result = ListNode(0)
        head = result

        carry = 0
        add = 0
        while l1 or l2:
            if l1:
                add += l1.val
                l1 = l1.next

            if l2:
                add += l2.val
                l2 = l2.next

            carry, add = divmod(add, 10)
            result.next = ListNode(add)
            result = result.next
            add = carry

        if carry > 0:
            result.next = ListNode(carry)

        return head.next


if __name__ == '__main__':
    l1 = ListNode(2)
    l1.next = ListNode(4)
    l1.next.next = ListNode(3)

    l2 = ListNode(5)
    l2.next = ListNode(6)
    l2.next.next = ListNode(4)

    result = Solution().addTwoNumbers(l1, l2)

    print(result.val)
    print(result.next.val)
    print(result.next.next.val)

File: src/main/leetcode/_002_addTwoNumbers/AddTwoNumbers.java
package leetcode._002_addTwoNumbers;

/*
 *   You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their
 *   nodes contain a single digit. Add the two numbers and return it as a linked list.
 *
 *
 *   You may assume the two numbers do not contain any leading zero, except the number 0 itself.
 *
 *   Example:
 *
 *   Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
 *   Output: 7 -> 0 -> 8
 *   Explanation: 342 + 465 = 807.
 */

import leetcode.data.ListNode;

/**
 * Created by luyao
 * on 2018/11/2 9:23
 */
public class AddTwoNumbers {

    public static ListNode solution(ListNode l1, ListNode l2) {
        ListNode result = new ListNode(0);
        ListNode head = result;
        int sum = 0, carry=0;
        while (l1 != null || l2 != null) {

            if (l1 != null) {
                sum += l1.val;
                l1 = l1.next;
            }

            if (l2 != null) {
                sum += l2.val;
                l2 = l2.next;
            }

            result.next = new ListNode(sum % 10);
            result = result.next;
            carry = sum / 10;
            sum = carry;
        }
        if (carry>0) result.next=new ListNode(carry);
        return head.next;
    }

}
